modules/sitemaps/sitemap-buffer.php

```diff
@@ -0,0 +1,202 @@
+ * the details of checking these constraints.
+ *
+ * This file defines the Jetpack_Sitemap_Buffer class, which provides
+ * an API for building and filling sitemaps.
+ *
+ * @since 4.6.0
+ * @package Jetpack
+ */
+
+/**
+ * A buffer for constructing sitemap xml files.
+ *
+ * Models a list of strings such that
+ *
+ * 1. the list must have a bounded number of entries,
+ * 2. the concatenation of the strings must have bounded
+ *      length (including some header and footer strings), and
+ * 3. each item has a timestamp, and we need to keep track
+ *      of the latest timestamp of the items in the list.
+ *
+ * @since 4.6.0
+ */
+class Jetpack_Sitemap_Buffer {
+	/**
+	 * Largest number of items the buffer can hold.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var int $item_capacity The item capacity.
+	 */
+	private $item_capacity;
+
+	/**
+	 * Largest number of bytes the buffer can hold.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var int $byte_capacity The byte capacity.
+	 */
+	private $byte_capacity;
+
+	/**
+	 * Footer text of the buffer; stored here so it can be appended when the buffer is full.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var string $footer_text The footer text.
+	 */
+	private $footer_text;
+
+	/**
+	 * The buffer contents.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var string The buffer contents.
+	 */
+	private $buffer;
+
+	/**
+	 * Flag which detects when the buffer is full.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var bool $is_full_flag The flag value. This flag is set to false on construction and only flipped to true if we've tried to add something and failed.
+	 */
+	private $is_full_flag;
+
+	/**
+	 * Flag which detects when the buffer is empty.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var bool $is_empty_flag The flag value. This flag is set to true on construction and only flipped to false if we've tried to add something and succeeded.
+	 */
+	private $is_empty_flag;
+
+	/**
+	 * The most recent timestamp seen by the buffer.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 * @var string $timestamp Must be in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	private $timestamp;
+
+	/**
+	 * Construct a new Jetpack_Sitemap_Buffer.
+	 *
+	 * @since 4.6.0
+	 *
+	 * @param int    $item_limit The maximum size of the buffer in items. Default is 50000.
+	 * @param int    $byte_limit The maximum size of the buffer in bytes. Default is 10485760 (10MB).
+	 * @param string $header The string to prepend to the entire buffer.
+	 * @param string $footer The string to append to the entire buffer.
+	 * @param string $time The initial timestamp of the buffer. Must be in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	public function __construct(
+		$item_limit = 50000,
+		$byte_limit = 10485760,
+		$header = '',
+		$footer = '',
+		$time
+	) {
+		$this->item_capacity = max( 1, intval( $item_limit ) );
+		$this->byte_capacity = $byte_limit - strlen( $header ) - strlen( $footer );
```

Comment:
The fix for this is [here](https://github.com/Automattic/jetpack/pull/5503/commits/290dde2a60c95a556c9b28a0385c9cad69ebd60a?diff=unified); I'd appreciate it if you could verify that I'm using `mbstring_binary_safe_encoding` correctly.